Change (and rename) Quaternion cubic_slerp to perform component-wise cubic interpolation

GNSS-Stylist opened the pull request:
Fixes #57951

- Function cubic_slerp changed to perform component-wise cubic interpolation and renamed to cubic_interpolate.
- Rewrote log and exp-functions based on https://users.aalto.fi/~ssarkka/pub/quat.pdf
- Added a new optional parameter to cubic_slerp that can be used to flip (or not) the quaternions to the shorter path. Default is to flip. "Base" quaternion is used as a reference (never flipped). This might be used for blending using pre-flipped (or "cached") quaternions in animation.

This breaks compatibility as cubic_slerp is renamed to cubic_interpolate.

Didn't update doc yet. I can try to do that if this pull request looks ok otherwise(?)

GNSS-Stylist committed:
Change (and rename) Quaternion cubic_slerp to perform component-wise cubic interpolation

Also:
- Rewrote log and exp-functions based on https://users.aalto.fi/~ssarkka/pub/quat.pdf
- Added a new optional parameter to cubic_slerp that can be used to flip (or not) the quaternions to the shorter path. Default is to flip. "Base" quaternion is used as a reference (never flipped).

fire commented:
The class reference needs to be updated and @TokageItLab said this pr will be checked.

TokageItLab reviewed (changes requested):
This PR is totally wrong. The calculation itself is wrong and the path and interpolation is completely crazy since it doesn't care flip problem. You can see it in the sample project in https://github.com/godotengine/godot/issues/57951.

The exp/log calculation is already implemented correctly and does not need to be changed, I say again that the only remaining problem is the sign flip issue https://github.com/godotengine/godot/issues/57951#issuecomment-1046425597. Even if we calculate instantaneous shorter paths, the sign changes when the interval changes, which breaks the path calculation, so we need to cache the sign of the first calculation in some ways.

GNSS-Stylist commented:
Sorry, I don't understand what's wrong as interpolation looks quite smooth here... Video from fire's fixed reproduction project ([cubic_slerp_test_duplicate_track.zip](https://github.com/godotengine/godot/files/8383420/cubic_slerp_test_duplicate_track.zip)) (earlier seemed to have keys almost at the same timestamp (2.9995 and 3s))  in https://github.com/godotengine/godot/issues/57951:

https://user-images.githubusercontent.com/50438441/180313647-20cec35b-b5fd-4f4c-b0ef-93ba03d92ff7.mp4

And from https://github.com/fire/godot-40592 :

https://user-images.githubusercontent.com/50438441/180315795-c8e025fa-890d-463c-8d8e-c4cd13cac472.mp4

https://user-images.githubusercontent.com/50438441/180318056-b9aa2ea8-e73d-42bd-af18-373272489356.mp4

This tries to take care of flipping, but only taking quaternions used in the interpolation itself into account. If I understood correctly the flipping problem is related to animation blending(?) I don't really know much about that, but isn't it a different problem? I added an optional parameter to the interpolation function to make it possible to avoid automatic flipping so it should be possible to use it in animation if so decided.

About blending though: I used kind of blending in one project where I "blended" between a quaternion (that was animated using the method in this PR) and another one formed from euler angles (because they felt more intuitive when operating in "FPS"-style coordinates with a camera). There I just used slerp to interpolate between those. But they were always quite close to one another, so maybe more generic blending causes different problems.

Exp&log-functions could indeed be leftovers. I just tried if the problems could be caused by them when kind of trying everything earlier. The document I stumbled upon approached the calculation from quite different angle and I just couldn't tell if they were identical. I can try to revert them if needed. However if this approach is "totally wrong", then it might not be necessary. ;)

TokageItLab commented:
> Sorry, I don't understand what's wrong as interpolation looks quite smooth here... Video from fire's fixed reproduction project ([cubic_slerp_test_duplicate_track.zip](https://github.com/godotengine/godot/files/8383420/cubic_slerp_test_duplicate_track.zip)) (earlier seemed to have keys almost at the same timestamp (2.9995 and 3s)) in https://github.com/godotengine/godot/issues/57951:

Oh, that was my bad. I am very sorry. I did not realize that. I'll look into it again later in more detail.

TokageItLab commented:
At least from the video I can see some points that seem to be broken. Transitions like the one at 9 seconds in the second video are wrong in the cubic_interpolate, which is probably a hitch caused by a sudden sign flip.

TokageItLab commented:
Ah, I remembered what we had worked out before on this.

This PR confused me, because it tries to interpolate quaternion values directly without using exp. That way could definitely take the shortest path, but the problem is that it is clearly not a slerp. So I remembered that you are trying to change the method name.

I and @fire were previously trying to solve this using exp map. However, we ran into a flip issue with exp map, so that issue has been left alone for now.

There are also other problems #60154 with cubic_interpolate that need to be fixed as well. So we will get back to this issue later, but if this PR-like approach is taken, it will be when we give up on the flip issue in the exp map.

GNSS-Stylist committed:
Revert Quaternion log and exp-functions

Made some tests and comparisons and these returned identical results. Therefore there's no need to modify the implementation.

From comment https://github.com/godotengine/godot/pull/63287#issuecomment-1191990751 :
Exp&log-functions could indeed be leftovers. I just tried if the problems could be caused by them when kind of trying everything earlier. The document I stumbled upon approached the calculation from quite different angle and I just couldn't tell if they were identical. I can try to revert them if needed.

GNSS-Stylist committed:
Fix flipping logic of Quaternion::cubic_interpolate()

There were edge cases causing sudden direction changes when dot-product between subsequent rotations was exactly 0 (every component being zero in either of the quaternions (or practically at least in this case 180 deg rotation between them around some axis)).

Left some test output in #if false-block so that the zero dot-products can be seen in godot's output. Will remove those if this change is accepted otherwise.

GNSS-Stylist commented:
> The exp/log calculation is already implemented correctly and does not need to be changed

You are right. Wrote some comparison/test code for these methods and results were identical, at least when using the quaternions used in the existing tests (which is quite a lot). Reverted the old implementations and made a commit.

> Transitions like the one at 9 seconds in the second video are wrong in the cubic_interpolate, which is probably a hitch caused by a sudden sign flip.

Investigated this a bit deeper and it was caused by quaternion dot-product between subsequent rotations being exactly 0 (every component being zero in either of the quaternions (or practically at least in this case 180 deg rotation between them around some axis)). I think I found a solution and committed a fix.

Not sure if this is related to your flipping problem with exp map (as I don't understand it well enough). But feel free to use it in your approach if you find it useful.

> This PR confused me, because it tries to interpolate quaternion values directly without using exp. That way could definitely take the shortest path, but the problem is that it is clearly not a slerp. So I remembered that you are trying to change the method name.

cubic_slerp felt quite wrong with this type of interpolation, so that's why I changed it. And you actually seemed to agree with the name change earlier if this approach to interpolation is taken (from https://github.com/godotengine/godot/issues/57951#issuecomment-1036844786 ) ;) :

> If I cannot find a way to resolve the kickback, it may look smoother to interpolate each element directly. Then, maybe we should avoid combining slerp() and cubic_interpolation() and we should choose to interpolate each element directly as you said, and rename the method Quaternion::cubic_slerp() to Quaternion::cubic_interpolate().

> There are also other problems https://github.com/godotengine/godot/issues/60154 with cubic_interpolate that need to be fixed as well. So we will get back to this issue later, but if this PR-like approach is taken, it will be when we give up on the flip issue in the exp map.

I have used some kind of "arbitrary interval" cubic hermite spline interpolation, based on https://en.wikipedia.org/wiki/Cubic_Hermite_spline on my own project. I will add a comment about that into the issue mentioned.

While twiddling with this I also made some modification into the test project https://github.com/fire/godot-40592 , PR: https://github.com/fire/godot-40592/pull/3 like added a bunch of rotations with dot-product 0 into the end of the animation.

fire commented on core/math/quaternion.cpp:
```diff
@@ -189,16 +195,70 @@ Quaternion Quaternion::slerpni(const Quaternion &p_to, const real_t &p_weight) c
+	real_t t2 = t * t;
+	real_t t3 = t2 * t;
+
+	return 0.5f *
```
I remember two attempts to consolidate the math equations. Can we check where they went and see if we're duplicating.

GNSS-Stylist replied:
This was indeed implemented in Math::cubic_interpolate (and animation was also using it to interpolate reals) so i changed this to use it instead. https://github.com/godotengine/godot/pull/63287/commits/4b425786c48c255be7c24200ff84eb475a3ac262

fire commented on core/math/quaternion.cpp:
```diff
@@ -189,16 +195,70 @@ Quaternion Quaternion::slerpni(const Quaternion &p_to, const real_t &p_weight) c
+		q3 = p_postq;
+	}
+
+	// TODO: Remove test output
```
Is this todo still valid?

GNSS-Stylist replied:
I'd like to remove the test output code completely (instead of disabling it using #if) finally, but left this for testing purposes for now. I will remove it if this method is decided to be merged.

fire commented:
The class reference also needs to be updated. 

> godot --doctool

GNSS-Stylist committed:
Replace local real_t interpolation function with Math::cubic_interpolate

This was already moved into a common module (core/math/math_funcs.h) during some refactoring so better to use it instead.

GNSS-Stylist committed:
Replace cubic_slerp with cubic_interpolate in Quaternion doc

Pondered a while how to phrase the description for flip_to_shortest_path-parameter. Now the description is quite long (and with an additional note even), so feel free to phrase it better.

fire commented:
Can you check https://github.com/godotengine/godot/pull/63380 ?

akien-mga commented:
Superseded by #63380.
